src/it/scala/com/wavesplatform/it/BalancesSpecification.scala

```diff
@@ -0,0 +1,318 @@
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+
+    first = result(sender.createAddress, 5.seconds)
+    second = result(sender.createAddress, 5.seconds)
+    third = result(sender.createAddress, 5.seconds)
+
+    val f = for {
+      fb <- sender.lastBlock
+      txs <- Future.sequence(Seq(
+        sender.transfer(richBalance, first, 200 * Constants.UnitsInWave, sender.fee(TransactionType.TransferTransaction)),
+        sender.transfer(richBalance, second, 100 * Constants.UnitsInWave, sender.fee(TransactionType.TransferTransaction)),
+        sender.transfer(richBalance, third, 100 * Constants.UnitsInWave, sender.fee(TransactionType.TransferTransaction)))
+      )
+      _ <- {
+        val allNodesAllTxsFindInBlockFutures = for {
+          tx <- txs
+          node <- allNodes
+        } yield {
+          node.findBlock(_.transactions.exists(_.id == tx.id), fb.height)
+        }
+        Future.sequence(allNodesAllTxsFindInBlockFutures)
+      }
+
+      r1 <- sender.balance(first).map(_.balance == 200 * Constants.UnitsInWave)
+      r2 <- sender.balance(second).map(_.balance == 100 * Constants.UnitsInWave)
+      r3 <- sender.balance(third).map(_.balance == 100 * Constants.UnitsInWave)
+    } yield r1 && r2 && r3
+    val res = result(f, 90 seconds)
+    res shouldBe true
+  }
+
+  private def assertBalances(acc: String, balance: Long, effectiveBalance: Long): Future[Unit] = {
+    for {
+      newBalance <- sender.balance(acc).map(_.balance)
+      newEffectiveBalance <- sender.effectiveBalance(acc).map(_.balance)
+    } yield {
+      newEffectiveBalance shouldBe effectiveBalance
+      newBalance shouldBe balance
+    }
+  }
+
+  private def assertAssetBalance(acc: String, assetIdString: String, balance: Long): Future[Unit] = {
+    sender.assetBalance(acc, assetIdString).map(_ shouldBe balance)
+  }
+
+  test("leasing is works correctly") {
+    val f = for {
+      fb <- sender.lastBlock
+
+      _ <- assertBalances(first, 200 * Constants.UnitsInWave, 200 * Constants.UnitsInWave)
+      _ <- assertBalances(second, 100 * Constants.UnitsInWave, 100 * Constants.UnitsInWave)
+
+      createdLeaseTxId <- sender.lease(first, second, 100 * Constants.UnitsInWave, fee = 10 * Constants.UnitsInWave).map(_.id)
+
+      _ <- Future.sequence(allNodes.map(_.findBlock(_.transactions.exists(_.id == createdLeaseTxId), fb.height)))
+
+      _ <- assertBalances(first, 190 * Constants.UnitsInWave, 90 * Constants.UnitsInWave)
+      _ <- assertBalances(second, 100 * Constants.UnitsInWave, 200 * Constants.UnitsInWave)
+    } yield succeed
+
+    result(f, 1 minute)
+  }
+
+  test("Сan not make leasing without having enough money") {
+    val f = for {
+      fb <- sender.lastBlock
+
+      _ <- assertBalances(first, 190 * Constants.UnitsInWave, 90 * Constants.UnitsInWave)
+      _ <- assertBalances(second, 100 * Constants.UnitsInWave, 200 * Constants.UnitsInWave)
+
+      leaseFailureAssertion <- recoverToSucceededIf[IOException] {
```

Comment:
Why is `recoverToSucceededIf` used here? All requests are supposed to be retried when an network-related exception is thrown.